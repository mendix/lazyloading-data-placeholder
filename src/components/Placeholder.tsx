import React from 'react';

import { Ghost } from './Ghost';

/**
 * `style` and `class` are default properties that are not defined in widget.config.ejs
 * You need to do some extra work to make them usable
 */

export interface PlaceholderProps {
  skeletonSVG: string;
  repeatCount: number;
  style: React.CSSProperties;
  className?: string;
  placeholderClassName?: string;
  itemClassName?: string;
  animate: boolean;
  animateInterval: number;
  ariaLabel: string;
  gradientRatio: number;
  width: number;
  height: number;
  preserveAspectRatio: string;
  primaryColor: string;
  primaryOpacity: number;
  secondaryColor: string;
  secondaryOpacity: number;
  rtl: boolean;
  speed: number;
}

const min = 20;

const getLength = (maxWidth: number) => {
  let w = 0;
  do {
    w = Math.round(Math.random() * 100);
  } while (w < min || w > maxWidth);
  return w;
};

const sanitizeSkeleton = (containerWidth: number, containerHeight: number) => {
  let topCorner = 10;
  let skeleton = '<svg>';
  do {
    let leftCorner = 10;
    do {
      let rectWidth = getLength(containerWidth - leftCorner - min);
      skeleton += `<rect x="${leftCorner}" y="${topCorner}" rx="4" ry="4" width="${rectWidth}" height="10"></rect>`;
      leftCorner += rectWidth + 20;
    } while (leftCorner < containerWidth - 2 * min);
    topCorner += 20;
  } while (topCorner < containerHeight - min);
  return skeleton + '</svg>';
};

export const Placeholder = (props: PlaceholderProps) => {
  let content = [];
  const childProps = {
    ...props,
    height: Math.round(props.height / props.repeatCount),
    skeletonSVG:
      props.skeletonSVG ||
      sanitizeSkeleton(props.width, props.height / props.repeatCount),
  };

  for (var i = 0; i < props.repeatCount; i++) {
    content.push(<Ghost key={i} {...childProps} />);
  }

  return (
    <div
      className={props.placeholderClassName}
      style={{
        display: 'flex',
        flexDirection: 'column',
        width: props.width,
        height: props.height,
      }}
    >
      {content}
    </div>
  );
};
