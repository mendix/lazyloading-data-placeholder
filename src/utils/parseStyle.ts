function getProperty(pair: string[]) {
  const name = pair[0]
    .trim()
    .replace(/(-.)/g, (match: string) => match[1].toUpperCase());
  return { [name]: pair[1].trim() };
}

/**
 *  Parse the added styles to the widget via the 'style' field in the modeler.
 * @public
 * @method
 * @param {string} style - styles as string
 * @return {Object} the resulted style object
 */
export default function parseStyle(style = '') {
  try {
    return style.split(';').reduce((styleObject, line) => {
      const pair = line.split(':');
      return pair.length === 2
        ? { ...styleObject, ...getProperty(pair) }
        : styleObject;
    }, {});
  } catch (error) {
    console.log('Failed to parse style', style, error); // eslint-disable-line no-console
  }
  return {};
}
