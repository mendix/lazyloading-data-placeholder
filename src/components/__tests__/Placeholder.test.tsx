import { Placeholder, PlaceholderProps } from '../Placeholder';

import React from 'react';
import parseStyle from '../../utils/parseStyle';
import { render } from '@testing-library/react';

describe('Placeholder Component Test Unit', () => {
  it('should render correctly', () => {
    const props: PlaceholderProps = {
      skeletonSVG: `<svg style="width:100%; height:100%;">
        <rect x="10%" y="10%" rx="4" ry="4" width="45%" height="10"></rect>
        <rect x="60%" y="10%" rx="4" ry="4" width="35%" height="10"></rect>
      </svg>`,
      repeatCount: 2,
      containerClassName: 'myContainerClass',
      containerStyle: parseStyle('width:400px;height:300px;'),
      itemClassName: 'myItemClass',
      itemStyle: parseStyle('height:20px;'),
    };
    const { container } = render(<Placeholder {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
