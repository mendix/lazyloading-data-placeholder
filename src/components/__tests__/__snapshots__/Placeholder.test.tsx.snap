// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Component Test Unit should render correctly 1`] = `
<div
  class="lazyloading-data-placeholder myContainerClass"
  style="height: 300px; width: 400px;"
>
  <div
    class="myItemClass"
    style="height: 20px;"
  >
    <svg
      style="width:100%; height:100%;"
    >
      
        
      <rect
        height="10"
        rx="4"
        ry="4"
        width="45%"
        x="10%"
        y="10%"
      />
      
        
      <rect
        height="10"
        rx="4"
        ry="4"
        width="35%"
        x="60%"
        y="10%"
      />
      
      
    </svg>
  </div>
  <div
    class="myItemClass"
    style="height: 20px;"
  >
    <svg
      style="width:100%; height:100%;"
    >
      
        
      <rect
        height="10"
        rx="4"
        ry="4"
        width="45%"
        x="10%"
        y="10%"
      />
      
        
      <rect
        height="10"
        rx="4"
        ry="4"
        width="35%"
        x="60%"
        y="10%"
      />
      
      
    </svg>
  </div>
</div>
`;
